sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/tuplesketchesAggregates.scala

```diff
@@ -0,0 +1,843 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.datasketches.tuple.{Intersection, Sketch, Summary, Union, UpdatableSketch, UpdatableSketchBuilder, UpdatableSummary}
+
+import org.apache.spark.SparkUnsupportedOperationException
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{ExpectsInputTypes, Expression, ExpressionDescription, Literal}
+import org.apache.spark.sql.catalyst.expressions.aggregate.TypedImperativeAggregate
+import org.apache.spark.sql.catalyst.trees.{QuaternaryLike, TernaryLike}
+import org.apache.spark.sql.catalyst.util.{ArrayData, CollationFactory, ThetaSketchUtils}
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.internal.types.StringTypeWithCollation
+import org.apache.spark.sql.types.{AbstractDataType, ArrayType, BinaryType, DataType, DoubleType, FloatType, IntegerType, LongType, StringType, StructType}
+import org.apache.spark.unsafe.types.UTF8String
+
+sealed trait TupleSketchState {
+  def serialize(): Array[Byte]
+  def eval(): Array[Byte]
+}
+case class UpdatableTupleSketchBuffer[U, S <: UpdatableSummary[U]](sketch: UpdatableSketch[U, S])
+    extends TupleSketchState {
+  override def serialize(): Array[Byte] = sketch.compact.toByteArray
+  override def eval(): Array[Byte] = sketch.compact.toByteArray
+}
+case class UnionTupleAggregationBuffer[S <: Summary](union: Union[S]) extends TupleSketchState {
+  override def serialize(): Array[Byte] = union.getResult.toByteArray
+  override def eval(): Array[Byte] = union.getResult.toByteArray
+}
+case class IntersectionTupleAggregationBuffer[S <: Summary](intersection: Intersection[S])
+    extends TupleSketchState {
+  override def serialize(): Array[Byte] = intersection.getResult.toByteArray
+  override def eval(): Array[Byte] = intersection.getResult.toByteArray
+}
+case class FinalizedTupleSketch[S <: Summary](sketch: Sketch[S]) extends TupleSketchState {
+  override def serialize(): Array[Byte] = sketch.toByteArray
+  override def eval(): Array[Byte] = sketch.toByteArray
+}
+
+/**
+ * The TupleSketchAgg function utilizes a Datasketches TupleSketch instance to count a
+ * probabilistic approximation of the number of unique values in a given column with associated
+ * summary values, and outputs the binary representation of the TupleSketch.
+ *
+ * See [[https://datasketches.apache.org/docs/Tuple/TupleSketches.html]] for more information.
+ *
+ * @param child
+ *   child expression (struct with key and summary value) against which unique counting will occur
```

Comment:
Where is the user expected to get both struct values from? In the example below, we have 

```
_FUNC_(struct(col, 1.0D), 12, 'double', 'sum')
```

Is the `1.0D` argument expected to be a constant most of the time, or where will it generally come from?

Reply:
Sounds good. I'd suggest we force the caller to provide the summary argument explicitly (rather than using 1 as the default) since other sketch functions could probably provide the same functionality (in a simpler way) for the case of a summary value of literal 1 for every input row.